R/sagemaker.R

```diff
@@ -0,0 +1,481 @@
+                                     docker_args = list(),
+                                     repo_name = glue("vetiver-sagemaker-{name}"),
+                                     compute_type = c(
+                                         "BUILD_GENERAL1_SMALL",
+                                         "BUILD_GENERAL1_MEDIUM",
+                                         "BUILD_GENERAL1_LARGE",
+                                         "BUILD_GENERAL1_2XLARGE"
+                                     ),
+                                     instance_type = "ml.t2.medium",
+                                     ...) {
+
+    if (!inherits(board, "pins_board_s3")) {
+        stop_input_type(board, "an S3 pins board")
+    }
+
+    repo_name <- ifelse(
+        grepl(":", repo_name),
+        repo_name,
+        glue("{repo_name}:{strftime(Sys.time(), '%Y-%m-%d')}")
+    )
+
+    # build image and push to aws ecr
+    image_uri <- vetiver_sm_build(
+        board = board,
+        name = name,
+        version = version,
+        predict_args = predict_args,
+        docker_args = docker_args,
+        repository = repo_name,
+        compute_type = compute_type,
+        bucket = board$bucket
+    )
+
+    args <- list2(...)
+    tags <- list_modify(args$tags, "vetiver")
+    args$tags <- NULL
+
+    # create sagemaker model
+    model_name <- vetiver_sm_model(image_uri, tags = tags)
+
+    # create sagemaker endpoint
+    endpoint_args <- c(
+        model_name = model_name,
+        instance_type = instance_type,
+        tags = tags,
+        args
+    )
+    endpoint <- do.call(vetiver_sm_endpoint, endpoint_args)
+    return(endpoint)
+}
+
+#' Deploy a vetiver model API to Amazon SageMaker with modular functions
+#'
+#' @description
+#' Use the function [vetiver_deploy_sagemaker()] for basic deployment on
+#' SageMaker, or these three functions together for more advanced use cases:
+#' - `vetiver_sm_build()` generates and builds a Docker image on SageMaker for
+#' a vetiver model
+#' - `vetiver_sm_model()` creates an Amazon SageMaker model
+#' - `vetiver_sm_endpoint()` deploys an Amazon SageMaker model endpoint
+#'
+#' @inheritParams vetiver_prepare_docker
+#' @param repository The ECR repository and tag for the image as a character.
+#' Defaults to `sagemaker-studio-${domain_id}:latest`.
+#' @param compute_type The [CodeBuild](https://aws.amazon.com/codebuild/)
+#' compute type as a character. Defaults to `BUILD_GENERAL1_SMALL`.
+#' @param role The IAM role name for CodeBuild to use as a character. Defaults
+#' to the SageMaker Studio execution role.
+#' @param bucket The S3 bucket to use for sending data to CodeBuild as a
+#' character. Defaults to the SageMaker SDK default bucket.
+#' @param vpc_id ID of the VPC that will host the CodeBuild project
+#' (such as `"vpc-05c09f91d48831c8c"`).
+#' @param subnet_ids List of subnet IDs for the CodeBuild project
+#' (such as `list("subnet-0b31f1863e9d31a67")`).
```

Comment:
Is this right @DyfanJones? i.e. folks should use this argument like `list("subnet-0b31f1863e9d31a67")`?

Reply:
Yes that is correct, it ultimately goes to this formatting:

```r
list(
      "vpcId" = list("vpc1"),
      "subnets" = list("subnet_ids1"),
      "securityGroupIds" = list("security_group_ids1")
)
```

Which mimics this behaviour: https://github.com/aws-samples/sagemaker-studio-image-build-cli/blob/master/sagemaker_studio_image_build/cli.py#L58-L67

Would it be worth changing vetiver's parameter to vpc_config to align with the other SageMaker functions so it is consistent  🤔

Reply:
If we did take the `vpc_config` parameter route it would looks something like:

```r
vetiver_sm_build <- function(board,
                             name,
                             version = NULL,
                             path = fs::dir_create(tempdir(), "vetiver"),
                             predict_args = list(),
                             docker_args = list(),
                             repository = NULL,
                             compute_type = c(
                                 "BUILD_GENERAL1_SMALL",
                                 "BUILD_GENERAL1_MEDIUM",
                                 "BUILD_GENERAL1_LARGE",
                                 "BUILD_GENERAL1_2XLARGE"
                             ),
                             role = NULL,
                             bucket = NULL,
                             vpc_config = list(),
                             log = TRUE,
                             ...) {
    check_installed("smdocker")
    compute_type <- arg_match(compute_type)

    # create dockerfile using
    # https://github.com/rocker-org/rocker-versioned2/pkgs/container/r-ver
    docker_args <- list_modify(docker_args, port = 8080)
    predict_args <- list_modify(predict_args, path = "/invocations")

    vetiver_prepare_docker(
        board = board,
        name = name,
        version = version,
        path = path,
        predict_args = predict_args,
        docker_args = docker_args
    )

    image_uri <- smdocker::sm_build(
        repository = repository,
        compute_type = compute_type,
        role = role,
        dir = path,
        bucket = bucket,
        vpc_id = vpc_config$vpcId,
        subnet_ids = vpc_config$Subnets %||% list(),
        security_group_ids = vpc_config$SecurityGroupIds %||% list(),
        log = log,
        ...
    )

    return(image_uri)
}
```

Reply:
I think we should keep it the same as what `smdocker::sm_build()` does, whatever that is. Do you think it would be better to change both the smdocker and vetiver args to be more like the other SageMaker functions?

Reply:
Happy to keep it as it is. `smdocker` currently mimics it's python counter part, if there was a strong desire to move to vpc_config I can look into adding it to the `smdocker` :)